teamvault/apps/secrets/utils.py

```diff
@@ -1,18 +1,30 @@
-from random import choice
-from string import ascii_letters, digits, punctuation
+import string
+from datetime import time
+from itertools import chain
+from random import choice, sample, seed
 
 from django.conf import settings
 from django.core.files.uploadhandler import MemoryFileUploadHandler, StopUpload
 
 
-def generate_password(length=12, alphanum=False):
-    """
-    Returns a password of the given length.
-    """
-    char_pool = ascii_letters + digits
-    if not alphanum:
-        char_pool += punctuation
-    return "".join(choice(char_pool) for i in range(length))
+def generate_password(length, digits, upper, lower, special):
+    seed(time())
```

Comment:
That does feel unsafe. You should never seed with something that is not cryptographically safe. It is also not impossible for an attacker to roughly re-construct the time a password was generated and thus significantly reduce the amount of password-attempts needed.

In general the python is *not* suited for security purposes, as it clearly stated in the docs in a big-fat-red box ;)
Better use the secrets-Module that gives some best practices on how to be used:
https://docs.python.org/3/library/secrets.html#recipes-and-best-practices